doctest/extensions/doctest_mpi.h

```diff
@@ -2,80 +2,104 @@
 
 #ifdef DOCTEST_CONFIG_IMPLEMENT
 
-#include "doctest/doctest.h"
+#include "doctest/extensions/mpi_sub_comm.h"
 #include "mpi_reporter.h"
-
-#else
-
-#include "mpi.h"
-#include <numeric>
-#include <vector>
-#include "doctest/doctest.h"
-#include <cassert>
-#include <string>
+#include <unordered_map>
 
 namespace doctest {
 
-inline
-int mpi_world_nb_procs() {
-  int n;
-  MPI_Comm_size(MPI_COMM_WORLD, &n);
-  return n;
+// Each time a MPI_TEST_CASE is executed on N procs,
+// we need a sub-communicator of N procs to execute it.
+// It is then registered here and can be re-used
+// by other tests that requires a sub-comm of the same size
+std::unordered_map<int,mpi_sub_comm> sub_comms_by_size;
+
+// Record if at least one MPI_TEST_CASE was registered "skipped"
+// because there is not enought procs to execute it
+int nb_test_cases_skipped_insufficient_procs = 0;
+
+// Can be safely called before MPI_Init()
+//   This is needed for MPI_TEST_CASE because we use doctest::skip()
+//   to prevent execution of tests where there is not enough procs,
+//   but doctest::skip() is called during test registration, that is, before main(), and hence before MPI_Init()
```

Comment:
Another usecase: `doctest_discover_tests` (used for registering the tests with [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html))

=> there the binary is also executed to list the test cases

(I made it work with MPI-tests for my stuff but didn't make a PR yet as it requires some cleanup etc.)

Reply:
Good to know. Do you want me to add this use case to the source comment ? Or you can do it with your PR ?